### ISSUE TITLE
Passphrase support for ssl client certificates

### ISSUE THREAD

**Issue by ysf** [weechat/weechat#115]
When using  irc.server.irc-host.ssl_xxx with an unencrypted file/key, everything works fine. But using an encrypted .pem-file, a passphrase to decrypt is needed but isn't read and used by weechat. At the moment only unencrypted .pem-files can be used.

**Comment by ysf** [comment]
Is this, by any chance, scheduled or maybe even resolved yet? I do not want to sound demanding, I'm just curious if it's on a todo-list or alike.

**Comment by flashcode** [comment]
Not scheduled yet.
It is not easy to "ask" something to the user in weechat (not possible now).
One solution would be to add an option in server with the passphrase, and evaluate this option, so the value could be stored encrypted in sec.conf file (the option would contain something like "${sec.data.freenode_ssl_passphrase}").
And then when connecting to the IRC server, nothing would be asked to the user.

**Comment by joaopizani** [comment]
The idea of making it an option in sec.conf seems good, I vote for it :)  A good compromise of security and ease-of-use.

**Comment by bendem** [comment]
would love to see passwords from irc.conf stored in sec.conf as well.

**Comment by flashcode** [comment]
@bendem: you can already do that, with the format I mentioned above. The options with passwords are evaluated, so the passwords themselves can be stored in sec.conf (since version 0.4.2).
See the user's guide for more info: https://weechat.org/doc/devel/user#secured_data

**Comment by bendem** [comment]
oh neat, ignore my comment then.

**Comment by ysf** [comment]
I second joaos vote, sounds like a viable solution to me too.

**Comment by joaopizani** [comment]
Regarding the passphrases for SSL private key of the client certificate, you could try checking the option first (freenode.ssl_passphrase), and only if the option does not exists (or is blank), then ask...

**Comment by TauPan** [comment]
I'm wondering if gnutls_x509_privkey_import2 could be used in irc-sasl.c (see http://www.gnutls.org/manual/html_node/Managing-encrypted-keys.html#Managing-encrypted-keys ) at least if a password from the configuration could be passed in.

However the documentation states that "Since GnuTLS 3.4.0 this function will utilize the PIN callbacks if any." and I wonder if that means that the client would block while gnutls asks for the passphrase.

**Comment by TauPan** [comment]
I have a workaround, which bears some improvement but already works for me:

` irc-local-proxy.sh`
```sh
#!/usr/bin/env bash
HOSTPORT="$1"
CERTFILE="$2"
FIFO_PREFIX="$3"

PASSPIPE=$FIFO_PREFIX/passpipe
IRCPIPE=$FIFO_PREFIX/ircpipe
set -e
umask 077
rm -f -- "$PASSPIPE" "$IRCPIPE"
mkfifo -- "$PASSPIPE" "$IRCPIPE"

(
    (echo "SETPROMPT Cert Password" ; echo GETPIN) \
        | pinentry \
        | sed -ne 's/^D //p' \
              > "$PASSPIPE"
    rm -f -- "$PASSPIPE"
) &

nc -l 127.0.0.1 6667 < $IRCPIPE \
    | exec openssl s_client -connect "$HOSTPORT" -cert "$CERTFILE" \
           -pass file:"$PASSPIPE" -no_ssl2 -quiet > $IRCPIPE

# Use: irc-local-proxy.sh irc.myserver.tld:6697 ~/secret/MyCert.pem /tmp
```

This will open a local irc server on port 6667 which forwards the connection via `netcat` and `openssl s_client`.

And then you can configure weechat, e.g. like this:
```
/server add forwarded localhost/6667
/alias add forwarded /exec -sh -name forward-proxy -nf -nosw irc-local-proxy.sh irc.myserver.tld:6697 $HOME'/secret/MyCert.pem' /tmp; /wait 3 /connect forwarded
```

And type '/forwarded' to connect. I'm working on a way to use pkcs11 and store the passphrase in weechat secure storage, but I have little time for this.

**Comment by acidvegas** [comment]
very nice approach @TauPan 
would love to see this fully implemented into Weechat @flashcode

**Comment by TauPan** [comment]
So I came up with the following rather hairy standalone version (without external scripts) assuming you have secure storage configured:
```sh
# Certificate File:
/eval /secure set IRC_TUNNEL_CERTFILE ${env:HOME}/secret/Friedrich Delgado.pem
# Host:Port of remote server
/secure set IRC_TUNNEL_DEST_HOSTPORT irc.mycorp.com:6697
# PREFIX for the FIFO Dateien. "pass" and "irc" are appended.
# The directory must exist beforehand.
/secure set IRC_TUNNEL_FIFO_PREFIX /tmp/irc_tunnel_fifo_
# Servername of server running on localhost
/secure set IRC_TUNNEL_LOCAL_SERVER localtunnelserver
# Define localtunnelserver. Only the name can be chosen freely at the
# moment, because the alias listens on localhost/6667
/server add localtunnelserver localhost/6667
# Passphrase for the cert in IRC_TUNNEL_CERTFILE
/secure set irc_tunnel_cert_passphrase TOTALLY SECRET PASSPHRASE 4a363913-855f-d0f4-b3fb-15f464daa748
# If you modify the alias, I recommend you start with a bogus
# passphrase here and make sure that it isn't echoed or logged
# anywhere.
#
# And the following monstrous alias:
/alias add irctunnel /eval /exec  -name cert-proxy -sh -nf -nosw rm -rf -- ${sec.data.IRC_TUNNEL_FIFO_PREFIX}{irc,pass} \; mkfifo -- ${sec.data.IRC_TUNNEL_FIFO_PREFIX}{irc,pass} \; nc -l 127.0.0.1 6667 < ${sec.data.IRC_TUNNEL_FIFO_PREFIX}irc| openssl s_client -connect "${sec.data.IRC_TUNNEL_DEST_HOSTPORT}" -cert "${sec.data.IRC_TUNNEL_CERTFILE}" -pass file:"${sec.data.IRC_TUNNEL_FIFO_PREFIX}pass" -no_ssl2 -quiet >  ${sec.data.IRC_TUNNEL_FIFO_PREFIX}irc; /eval /mute /wait 100ms /exec -norc -sh echo ${sec.data.irc_tunnel_cert_passphrase} > ${sec.data.IRC_TUNNEL_FIFO_PREFIX}pass; /wait 200ms /eval /connect ${sec.data.IRC_TUNNEL_LOCAL_SERVER}; /wait 300ms /eval exec -bg -sh rm -rf -- ${sec.data.IRC_TUNNEL_FIFO_PREFIX}pass
```

Same WARNING as before applies: This will allow anybody who can connect to port 6667 on your machine to circumvent authentication. So you should only use this on a box where nobody has remote access except yourself!

**Comment by sim642** [comment]
In case GitHub doesn't notify users subscribed to this issue about PRs referencing this, I'll mention that I've put together a small PR that hopefully adds this long-awaited capability: https://github.com/weechat/weechat/pull/1416. I invite everyone to try and test it out.

**Comment by TauPan** [comment]
This is in 2.7 now! \o/

**Comment by tjkirch** [comment]
Does this (as implemented in #1416) mean that GnuTLS 3+ is now required?  I can't build WeeChat 2.7 on my distro because of the change from `gnutls_x509_privkey_import` to `gnutls_x509_privkey_import2`, which was introduced in GnuTLS 3.0.21.

**Comment by sim642** [comment]
Is that really the case? I checked the [GnuTLS manual for `gnutls_x509_privkey_import2`](https://gnutls.org/manual/gnutls.html#index-gnutls_005fx509_005fprivkey_005fimport2-1) and there is no "**Since:** …" information for it, unlike many other functions have such information. By the lack of that information I assumed it has always been there.

If that's really the case, then WeeChat's GnuTLS version requirement doesn't need to be increased altogether. An additional `LIBGNUTLS_VERSION_NUMBER` macro check could be added to make WeeChat still compile with older GnuTLS versions but simply lack the support for SSL client certificate passwords, as it was before.

**Comment by flashcode** [comment]
@tjkirch: which distro still uses GnuTLS < 3.0.21, which was released in July 2012, so more than 7 years ago?

**Comment by tjkirch** [comment]
> Is that really the case?

@sim642 I looked at the git history for GnuTLS and found that `gnutls_3_0_21` was the first tag with `43d5826b171937b9415501a5e227c8ec1d24bc71` which introduced that function.  It might just be a missing "Since:" marker?

> WeeChat's GnuTLS version requirement doesn't need to be increased altogether

That would be great!

> which distro still uses GnuTLS < 3.0.21

@flashcode RHEL/CentOS 6 and its derivatives.  We don't all have choice over what we're allowed to use, but until now I've been able to use the latest WeeChat, so I was sad about this particular change.
